template/src/features/counter/counterSlice.js

```diff
@@ -24,16 +24,24 @@ export const counterSlice = createSlice({
 
 export const { increment, decrement, incrementByAmount } = counterSlice.actions;
 
-// The function below is called a thunk and allows us to perform async logic. It
-// can be dispatched like a regular action: `dispatch(incrementAsync(10))`. This
-// will call the thunk with the `dispatch` function as the first argument. Async
-// code can then be executed and other actions can be dispatched
-export const incrementAsync = amount => dispatch => {
-  setTimeout(() => {
-    dispatch(incrementByAmount(amount));
-  }, 1000);
+const fakeAsyncCall = async amount => {
+  await new Promise(resolve => setTimeout(resolve, 1000));
+  return amount;
 };
 
+// The function below is called a thunk and allows us to perform async logic. It
+// can be dispatched like a regular action: `dispatch(incrementAsync(10))`.
+// The `thunkApi` argument has `dispatch`, `getState` and other functions available.
+// See https://redux-toolkit.js.org/api/createAsyncThunk
+// Async code can then be executed and other actions can be dispatched.
+export const incrementAsync = createAsyncThunk(
+  'counter/incrementAsync',
+  async (amount, thunkApi) => {
+    const value = await fakeAsyncCall(amount);
+    thunkApi.dispatch(incrementByAmount(value));
```

Comment:
I think this code is a little less complicated but get's the same point across:
```suggestion
    await delay(1000);
    thunkApi.dispatch(incrementByAmount(amount));
```

Reply:
My main concern here is that the point of `createAsyncThunk` is to auto-dispatch actions based on a promise's lifecycle, and then handle those actions in a reducer, with the return value of this callback becoming the promise.

The hand-written example thunk wasn't making any AJAX calls, it was just showing you can have any kind of async logic at all, and have access to `getState`.

This use of `createAsyncThunk` doesn't fit either approach. It's not returning a payload, it's dispatching a separate action type, and it's not the right semantic use for `createAsyncThunk`.

I've got mixed opinions on including examples of `createAsyncThunk` right in the template. Biggest issue is that we don't have any API to call here.

Tagging @phryneas and @msutkowski to get their thoughts.

Reply:
I fully agree with these points and don't have any counters to them.

If you choose to keep the `createAsyncThunk` usage, I'd return the value and handle the case in `extraReducers` so that it more closely matches normal usage. I'd still question that implementation though.

Reply:
So I hadn't heard of `createAsyncThunk` when I made this comment. Reading up on it a bit I feel like this isn't the best showcase of `createAsyncThunk`. If the typical use-case is to make an api call, it should probably make an api call.  Otherwise people will be confused why we're using `createAsyncThunk` here.

Reply:
Yeah. Adding example functionality that actually queries something (mocked by `msw` or similar)  might be an idea - but here, `createAsyncThunk` is not used correctly.

Reply:
@markerikson I agree with what you're saying here, `createAsyncThunk` is not used as it would be normally. However I really like `createAsyncThunk` and wonder if it would be useful to have it used in the template.

I think there are a few options & I'm happy to do whichever you're most happy with.

* Revert the change here, don't use `createAsyncThunk`
  * I couldn't work out how to test the function that was there before at all - any tips?
  * or don't test that function
* Use  `createAsyncThunk` and add a reducer in `extraReducers` for handling `incrementAsync/resolved`
  * Use the `fakeAsyncCall` that I've added
  * Or do more elaborate api mocking

Thoughts?

Reply:
@markerikson I've pushed an update that uses `createAsyncThunk` more appropriately. What do you think?